diff --git a/src/gui/configure.cmake b/src/gui/configure.cmake
index 7f6c48fb..a3b8e4de 100644
--- a/src/gui/configure.cmake
+++ b/src/gui/configure.cmake
@@ -1101,7 +1101,7 @@ qt_feature("egl_x11" PRIVATE
 qt_feature("eglfs" PRIVATE
     SECTION "Platform plugins"
     LABEL "EGLFS"
-    CONDITION NOT ANDROID AND NOT APPLE AND NOT WIN32 AND NOT WASM AND NOT QNX AND QT_FEATURE_egl
+    CONDITION NOT ANDROID AND NOT APPLE AND NOT WIN32 AND NOT WASM AND QT_FEATURE_egl
 )
 qt_feature("eglfs_brcm" PRIVATE
     LABEL "EGLFS Raspberry Pi"
@@ -1320,7 +1320,7 @@ qt_feature("clipboard" PUBLIC
     SECTION "Kernel"
     LABEL "QClipboard"
     PURPOSE "Provides cut and paste operations."
-    CONDITION NOT INTEGRITY AND NOT QNX AND NOT rtems
+    CONDITION NOT INTEGRITY AND NOT rtems
 )
 qt_feature_definition("clipboard" "QT_NO_CLIPBOARD" NEGATE VALUE "1")
 qt_feature("wheelevent" PUBLIC
diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
index 05ffb3b2..3ab66ad9 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms/qeglfskmsgbmintegration.cpp
@@ -44,7 +44,7 @@ EGLDisplay QEglFSKmsGbmIntegration::createDisplay(EGLNativeDisplayType nativeDis
     }
 
     if (getPlatformDisplay) {
-        display = getPlatformDisplay(EGL_PLATFORM_GBM_KHR, nativeDisplay, nullptr);
+        display = getPlatformDisplay(EGL_PLATFORM_GBM_KHR, reinterpret_cast<void*>(nativeDisplay), nullptr);
     } else {
         qCDebug(qLcEglfsKmsDebug, "No eglGetPlatformDisplay for GBM, falling back to eglGetDisplay");
         display = eglGetDisplay(nativeDisplay);
diff --git a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldeviceintegration.cpp b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldeviceintegration.cpp
index ece19f46..24350857 100644
--- a/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldeviceintegration.cpp
+++ b/src/plugins/platforms/eglfs/deviceintegration/eglfs_kms_egldevice/qeglfskmsegldeviceintegration.cpp
@@ -42,7 +42,7 @@ EGLDisplay QEglFSKmsEglDeviceIntegration::createDisplay(EGLNativeDisplayType nat
 
     const EGLint attribs[] = { EGL_DRM_MASTER_FD_EXT, egldevice_fd, EGL_NONE };
     if (m_funcs->has_egl_platform_device) {
-        display = m_funcs->get_platform_display(EGL_PLATFORM_DEVICE_EXT, nativeDisplay, attribs);
+        display = m_funcs->get_platform_display(EGL_PLATFORM_DEVICE_EXT, reinterpret_cast<void*>(nativeDisplay), attribs);
     } else {
         qWarning("EGL_EXT_platform_device not available, falling back to legacy path!");
         display = eglGetDisplay(nativeDisplay);
diff --git a/src/plugins/platforms/qnx/CMakeLists.txt b/src/plugins/platforms/qnx/CMakeLists.txt
index 0f9deaa0..c5c02b39 100644
--- a/src/plugins/platforms/qnx/CMakeLists.txt
+++ b/src/plugins/platforms/qnx/CMakeLists.txt
@@ -66,11 +66,9 @@ qt_internal_extend_target(QQnxIntegrationPlugin CONDITION QT_FEATURE_qqnx_pps
         PPS::PPS
 )
 
-qt_internal_extend_target(QQnxIntegrationPlugin CONDITION QT_FEATURE_clipboard AND QT_FEATURE_qqnx_pps
+qt_internal_extend_target(QQnxIntegrationPlugin CONDITION QT_FEATURE_clipboard
     SOURCES
         qqnxclipboard.cpp qqnxclipboard.h
-    LIBRARIES
-        clipboard
 )
 
 qt_internal_extend_target(QQnxIntegrationPlugin CONDITION QT_FEATURE_qqnx_imf AND QT_FEATURE_qqnx_pps
diff --git a/src/plugins/platforms/qnx/qqnxclipboard.cpp b/src/plugins/platforms/qnx/qqnxclipboard.cpp
index 3f27ec80..48a5d1ca 100644
--- a/src/plugins/platforms/qnx/qqnxclipboard.cpp
+++ b/src/plugins/platforms/qnx/qqnxclipboard.cpp
@@ -1,4 +1,5 @@
 // Copyright (C) 2011 - 2012 Research In Motion
+// Copyright (c) 2020 BlackBerry Limited
 // SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
 
 #undef QT_NO_FOREACH // this file contains unported legacy Q_FOREACH uses
@@ -8,36 +9,25 @@
 #include "qqnxclipboard.h"
 
 #include <QtGui/QColor>
+#include <QtGui/QImage>
 
+#include <QtCore/QBuffer>
 #include <QtCore/QDebug>
+#include <QtCore/QDirIterator>
 #include <QtCore/QMimeData>
+#include <QtCore/QSaveFile>
+#include <QtCore/QSet>
 #include <QtCore/QStringList>
 #include <QtCore/QUrl>
 
-#include <clipboard/clipboard.h>
+
 #include <errno.h>
 
 QT_BEGIN_NAMESPACE
 
 Q_LOGGING_CATEGORY(lcQpaClipboard, "qt.qpa.clipboard");
 
-// null terminated array
-static const char *typeList[] = {"text/html", "text/plain", "image/png", "image/jpeg", "application/x-color", 0};
-
-static QByteArray readClipboardBuff(const char *type)
-{
-    char *pbuffer;
-    if (is_clipboard_format_present(type) == 0) {
-        int size = get_clipboard_data(type, &pbuffer);
-        if (size != -1 && pbuffer) {
-            const QByteArray result = QByteArray(pbuffer, size);
-            free(pbuffer);
-            return result;
-        }
-    }
-
-    return QByteArray();
-}
+using namespace Qt::StringLiterals;
 
 class QQnxClipboard::MimeData : public QMimeData
 {
@@ -46,18 +36,18 @@ public:
     MimeData(QQnxClipboard *clipboard)
         : QMimeData(),
           m_clipboard(clipboard),
-          m_userMimeData(0)
+          m_userMimeData(nullptr),
+          m_formatsToCheck({u"text/html"_s,
+                            u"text/plain"_s,
+                            u"image/png"_s,
+                            u"image/jpeg"_s,
+                            u"application/x-color"_s})
     {
         Q_ASSERT(clipboard);
-
-        for (int i = 0; typeList[i] != 0; ++i) {
-            m_formatsToCheck << QString::fromUtf8(typeList[i]);
-        }
     }
 
     ~MimeData()
     {
-        delete m_userMimeData;
     }
 
     void addFormatToCheck(const QString &format) {
@@ -67,7 +57,7 @@ public:
 
     bool hasFormat(const QString &mimetype) const override
     {
-        const bool result = is_clipboard_format_present(mimetype.toUtf8().constData()) == 0;
+        const bool result = m_clipboard->hasFormat(mimetype);
         qCDebug(lcQpaClipboard) << "mimetype=" << mimetype << "result=" << result;
         return result;
     }
@@ -76,8 +66,8 @@ public:
     {
         QStringList result;
 
-        Q_FOREACH (const QString &format, m_formatsToCheck) {
-            if (is_clipboard_format_present(format.toUtf8().constData()) == 0)
+        for (const QString &format : m_formatsToCheck) {
+            if (m_clipboard->hasFormat(format))
                 result << format;
         }
 
@@ -87,8 +77,7 @@ public:
 
     void setUserMimeData(QMimeData *userMimeData)
     {
-        delete m_userMimeData;
-        m_userMimeData = userMimeData;
+        m_userMimeData.reset(userMimeData);
 
         // system clipboard API doesn't allow detection of changes by other applications
         // simulate an owner change through delayed invocation
@@ -99,17 +88,26 @@ public:
 
     QMimeData *userMimeData()
     {
-        return m_userMimeData;
+        return m_userMimeData.get();
     }
 
 protected:
     QVariant retrieveData(const QString &mimetype, QMetaType preferredType) const override
     {
-        qCDebug(lcQpaClipboard) << "mimetype=" << mimetype << "preferredType=" << preferredType.name();
-        if (is_clipboard_format_present(mimetype.toUtf8().constData()) != 0)
+        qCDebug(lcQpaClipboard) << "mimetype=" << mimetype << "preferredType=" << preferredType;
+        if (!m_clipboard->hasFormat(mimetype))
             return QMimeData::retrieveData(mimetype, preferredType);
 
-        const QByteArray data = readClipboardBuff(mimetype.toUtf8().constData());
+        const QByteArray data = m_clipboard->read(mimetype);
+        if (mimetype == "application/x-qt-image") {
+            QBuffer buffer;
+            buffer.setData(data);
+            buffer.open(QIODevice::ReadOnly);
+            QImage image;
+            if (image.load(&buffer, "PNG")) {
+                return QVariant::fromValue(image);
+            }
+        }
         return QVariant::fromValue(data);
     }
 
@@ -118,8 +116,7 @@ private Q_SLOTS:
     {
         if (m_userMimeData) {
             qCDebug(lcQpaClipboard) << "user data formats=" << m_userMimeData->formats() << "system formats=" << formats();
-            delete m_userMimeData;
-            m_userMimeData = 0;
+            m_userMimeData.reset();
             m_clipboard->emitChanged(QClipboard::Clipboard);
         }
     }
@@ -128,12 +125,20 @@ private:
     QQnxClipboard * const m_clipboard;
 
     QSet<QString> m_formatsToCheck;
-    QMimeData *m_userMimeData;
+    std::unique_ptr<QMimeData> m_userMimeData;
 };
 
 QQnxClipboard::QQnxClipboard()
-    : m_mimeData(new MimeData(this))
+    : m_mimeData(new MimeData(this)), m_serverAvailable(false)
 {
+    // The clipboard is only implemented if the desktop option is specified
+    // and if the QQNX_CLIPBOARD environment variable points to a directory
+    // that contains a .qnxclipboard node.
+    if (qEnvironmentVariableIsSet("QQNX_CLIPBOARD")) {
+        m_clipboardDir = QDir(qEnvironmentVariable("QQNX_CLIPBOARD"));
+        if (m_clipboardDir.exists(".qnxclipboard"))
+            m_serverAvailable = true;
+    }
 }
 
 QQnxClipboard::~QQnxClipboard()
@@ -152,12 +157,12 @@ void QQnxClipboard::setMimeData(QMimeData *data, QClipboard::Mode mode)
     if (m_mimeData->userMimeData() && m_mimeData->userMimeData() == data)
         return;
 
-    empty_clipboard();
+    clear();
 
     m_mimeData->clear();
     m_mimeData->setUserMimeData(data);
 
-    if (data == 0) {
+    if (!data) {
         emitChanged(QClipboard::Clipboard);
         return;
     }
@@ -165,13 +170,23 @@ void QQnxClipboard::setMimeData(QMimeData *data, QClipboard::Mode mode)
     const QStringList formats = data->formats();
     qCDebug(lcQpaClipboard) << "formats=" << formats;
 
-    Q_FOREACH (const QString &format, formats) {
-        const QByteArray buf = data->data(format);
+    for (const QString &format : formats) {
+        QByteArray buf;
+
+        // Handling for image data
+        if (format == "application/x-qt-image") {
+            QBuffer buffer(&buf);
+            buffer.open(QIODevice::WriteOnly);
+            QImage image = qvariant_cast<QImage>(data->imageData());
+            image.save(&buffer, "PNG");
+        } else {
+            buf = data->data(format);
+        }
 
         if (buf.isEmpty())
             continue;
 
-        int ret = set_clipboard_data(format.toUtf8().data(), buf.size(), buf.data());
+        bool ret = write(format, buf);
         qCDebug(lcQpaClipboard) << "set " << format << "to clipboard, size=" << buf.size() << ";ret=" << ret;
         if (ret)
             m_mimeData->addFormatToCheck(format);
@@ -193,6 +208,73 @@ QMimeData *QQnxClipboard::mimeData(QClipboard::Mode mode)
     return m_mimeData;
 }
 
+QFile QQnxClipboard::fileforFormat(const QString &format) const
+{
+    // If using a standard file system directory, format names of the standard
+    // "foo/bar" style cannot be handled properly.
+    QString fileName = format;
+    fileName.replace('/', '@');
+    return QFile(m_clipboardDir.filePath(fileName));
+}
+
+bool QQnxClipboard::hasFormat(const QString &format) const
+{
+    if (!m_serverAvailable)
+        return false;
+
+    QFile file = fileforFormat(format);
+    return file.exists() && file.size() > 0;
+}
+
+QByteArray QQnxClipboard::read(const QString &format) const
+{
+    if (!m_serverAvailable)
+        return QByteArray();
+
+    QFile file = fileforFormat(format);
+    if (!file.open(QIODevice::ReadOnly))
+        return QByteArray();
+
+    return file.readAll();
+}
+
+bool QQnxClipboard::write(const QString &format, const QByteArray &data) const
+{
+    if (!m_serverAvailable)
+        return false;
+
+    QFile file = fileforFormat(format);
+    if (!file.open(QIODevice::WriteOnly))
+        return false;
+
+    return file.write(data) == data.size();
+}
+
+void QQnxClipboard::clear()
+{
+    if (!m_serverAvailable)
+        return;
+
+    // Warning!
+    // Deletes all files in the given directory. Make sure QQNX_CLIPBOARD is set
+    // to the mount point of a clipboard service or to a dedicated directory.
+    // This is double-checked in the constructor, which will not start the
+    // clipboard unless it points to a directory that contains a ".qnxclipboard"
+    // node.
+    m_clipboardDir.refresh();
+    QDirIterator itr(m_clipboardDir);
+    while (itr.hasNext()) {
+        itr.next();
+        QString format = itr.fileName();
+        if (m_clipboardDir.exists(format)) {
+            qCDebug(lcQpaClipboard) << "deleting" << format;
+            bool removed = m_clipboardDir.remove(format);
+            if (!removed)
+                qCDebug(lcQpaClipboard) << "Failed to remove" << format;
+        }
+    }
+}
+
 QT_END_NAMESPACE
 
 #include "qqnxclipboard.moc"
diff --git a/src/plugins/platforms/qnx/qqnxclipboard.h b/src/plugins/platforms/qnx/qqnxclipboard.h
index 66f862f0..ee51e520 100644
--- a/src/plugins/platforms/qnx/qqnxclipboard.h
+++ b/src/plugins/platforms/qnx/qqnxclipboard.h
@@ -1,10 +1,13 @@
 // Copyright (C) 2011 - 2012 Research In Motion
+// Copyright (c) 2020 BlackBerry Limited
 // SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
 
 #ifndef QQNXCLIPBOARD_H
 #define QQNXCLIPBOARD_H
 
 #include <QtCore/qglobal.h>
+#include <QtCore/qdir.h>
+#include <QtCore/qloggingcategory.h>
 
 #if !defined(QT_NO_CLIPBOARD)
 #include <qpa/qplatformclipboard.h>
@@ -24,6 +27,14 @@ public:
 private:
     class MimeData;
     MimeData *m_mimeData;
+    QDir m_clipboardDir;
+    bool m_serverAvailable;
+
+    [[nodiscard]] QFile fileforFormat(const QString &format) const;
+    [[nodiscard]] bool hasFormat(const QString &format) const;
+    [[nodiscard]] QByteArray read(const QString &format) const;
+    [[nodiscard]] bool write(const QString &format, const QByteArray &data) const;
+    void clear();
 };
 
 QT_END_NAMESPACE
diff --git a/src/plugins/platforms/qnx/qqnxintegration.cpp b/src/plugins/platforms/qnx/qqnxintegration.cpp
index 1a60bf06..216bbe0a 100644
--- a/src/plugins/platforms/qnx/qqnxintegration.cpp
+++ b/src/plugins/platforms/qnx/qqnxintegration.cpp
@@ -31,7 +31,6 @@
 
 #if QT_CONFIG(qqnx_pps)
 #  include "qqnxbuttoneventnotifier.h"
-#  include "qqnxclipboard.h"
 #endif
 
 #if QT_CONFIG(qqnx_imf)
@@ -40,6 +39,10 @@
 #  include "qqnxinputcontext_noimf.h"
 #endif
 
+#if !defined(QT_NO_CLIPBOARD)
+#  include "qqnxclipboard.h"
+#endif
+
 #include <qpa/qplatforminputcontextfactory_p.h>
 #include <qpa/qplatforminputcontext.h>
 
@@ -446,10 +449,9 @@ QPlatformClipboard *QQnxIntegration::clipboard() const
 {
     qCDebug(lcQpaQnx) << Q_FUNC_INFO;
 
-#if QT_CONFIG(qqnx_pps)
     if (!m_clipboard)
         m_clipboard = new QQnxClipboard;
-#endif
+    
     return m_clipboard;
 }
 #endif
